Feature - gRPC subscription of minimalConsensusInfo (van->orc)

mxmar opened the pull request:
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

We need to switch to the gRPC server in `vanguard` side (`minimalConsensusInfo`)

**Which issues(s) does this PR fix?**

Fixes https://github.com/lukso-network/vanguard-consensus-engine/issues/14

**Other notes for review**

None

mxmar committed:
Feature: 11 grpc van sub
- added beacon (blockchain) implementation;
- added grpc server file;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Merge branch 'release/vanguard' into feature/11/grpc-van-sub

mxmar committed:
Feature: 11 grpc van sub
- added vanguard-apis for minimalConsensusInfo;
- added backend for minimalConsensusInfo with new apis;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- modified backend for minimalConsensusInfo;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- modified backend and grpc server for minimalConsensusInfo;
- still Send stream not working - need to modify apis for range infos;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- modified event data;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- modified backend and unit tests for minimalConsensusInfo;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- test for beacon-chain/blockchain for minimalConsensusInfo grpc changes;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- new apis mod;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- implementation done;
- starting testing phase;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- fixed bazel build file during testing;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- fixed bazel build file during testing;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- fixed bazel build file and deps file;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- delete bazel build for van_mock;

Signed-off-by: mxmar <macht4tech@gmail.com>

atif-konasl commented on beacon-chain/rpc/beacon/van_minimal_consensus_info.go:
```diff
@@ -0,0 +1,56 @@
+)
+
+// StreamMinimalConsensusInfo to orchestrator client every single time an unconfirmed block is received by the beacon node.
+func (bs *Server) StreamMinimalConsensusInfo(
```
Every time subscriber gets consensus info from epoch 0 which should not be happened. This event api should take `fromEpoch` from subscriber and send consensus infos from `fromEpoch`

mxmar replied:
Ready. I've provided this argument as MinimalConsensusRequest in protobuf + inside consensus engine as `ethpb.MinimalConsensusRequest`

mxmar committed:
Feature: 11 grpc van sub
- delete bazel build deps for old rpc;

Signed-off-by: mxmar <macht4tech@gmail.com>

Commit:
Generated mock file

mxmar committed:
Feature: 11 grpc van sub
- build is working for beacon-chain;

Signed-off-by: mxmar <macht4tech@gmail.com>

Commit:
Added bazel file in van_mock package

atif-konasl commented on beacon-chain/rpc/beacon/van_minimal_consensus_info.go:
```diff
@@ -0,0 +1,56 @@
+
+	for {
+		select {
+		case minConsensusEvent := <-minimalConsensusChannel:
```
From which point, this `minimalConsensusChannel` channel gets triggered. I could not find trigger point.

mxmar replied:
Refactored to use epochTicker instead of events.

atif-konasl commented on beacon-chain/rpc/beacon/van_minimal_consensus_info.go:
```diff
@@ -0,0 +1,56 @@
+)
+
+// StreamMinimalConsensusInfo to orchestrator client every single time an unconfirmed block is received by the beacon node.
+func (bs *Server) StreamMinimalConsensusInfo(
```
@mxmar If any block is skipped in any slot then what will happen? In that case, this method does not send consensus info to orchestrator. Right? Before starting this feature implementation, i suggested you to follow `StreamDuties(req *ethpb.DutiesRequest, stream ethpb.BeaconNodeValidator_StreamDutiesServer)` api from `rpc/validator/assignments.go`.

mxmar replied:
Refactored to epochTicker approach instead of events. Right now I'm checking `slot == endSlot` before sending consensus info.

mxmar committed:
Feature: 11 grpc van sub
- missing only the latest vanguard-apis;

Signed-off-by: mxmar <macht4tech@gmail.com>

atif-konasl commented on beacon-chain/rpc/beacon/van_minimal_consensus_info.go:
```diff
@@ -0,0 +1,67 @@
+	}
+
+	secondsPerEpoch := params.BeaconConfig().SecondsPerSlot * uint64(params.BeaconConfig().SlotsPerEpoch)
+	epochTicker := slotutil.NewSlotTicker(bs.GenesisTimeFetcher.GenesisTime(), secondsPerEpoch)
```
@mxmar I think it triggers at first slot of each epoch. But we need to trigger after last slot of each epoch. So i guess  you should write test case and need to check it.

mxmar replied:
I will try to figure out this now.

mxmar replied:
I've wrote go test -> `TestServer_StreamMinimalConsensusInfo`

atif-konasl commented on beacon-chain/rpc/beacon/van_minimal_consensus_info.go:
```diff
@@ -0,0 +1,67 @@
+			if err != nil {
+				return status.Errorf(codes.Unavailable, "Could not retrieve end slot of epoch: %v, err: %v", err, types.Epoch(slot))
+			}
+			if slot == endSlot {
```
I doubt that, it may not enter into if clause, because `epochTicker` will tick on first slot of each epoch.

mxmar replied:
I've removed this logic.

atif-konasl commented on beacon-chain/blockchain/van_minimal_consensus_info.go:
```diff
@@ -0,0 +1,197 @@
+package blockchain
```
In future, we will move this file to `core/helpers` package because it does not related with blockchain package

mxmar replied:
In my opinion we should move back to the `rpc/beacon` because `helpers` are more like `committees.go` logic than any of helpers. Maybe other solution is to have `core/consensus` like `core/validators` is. What do you think @atif-konasl ?

mxmar replied:
Done. As we have agreed on I've moved to the rpc/beacon package.

atif-konasl commented on beacon-chain/blockchain/van_minimal_consensus_info_test.go:
```diff
@@ -0,0 +1,128 @@
+package blockchain
```
In future, we will move this file to `core/helpers` package because it does not related with blockchain package

mxmar replied:
Done. As we have agreed on I've moved to the `rpc/beacon` package.

atif-konasl commented on beacon-chain/blockchain/van_minimal_consensus_info.go:
```diff
@@ -0,0 +1,197 @@
+	consensusInfos = append(consensusInfos, consensusInfo)
+	tempEpochIndex := consensusInfo.Epoch
+
+	for {
```
you can retrieve `currentEpoch := helpers.SlotToEpoch(bs.GenesisTimeFetcher.CurrentSlot())` and then iterate from `fromEpoch` to `currentEpoch` and prepare `consensusInfo` for each of the slot. I would not prefer using infinite loop in here.

mxmar replied:
As we agreed on, we're leaving infinity loop because of gRPC streaming - need to have this loop to handle stream/connection.

atif-konasl commented on beacon-chain/blockchain/van_minimal_consensus_info.go:
```diff
@@ -0,0 +1,197 @@
+
+//MinimalConsensusInfoRange provides minimalConsensusInfo for all epochs from epoch provided as an argument
+func (s *Service) MinimalConsensusInfoRange(
+	fromEpoch types.Epoch,
```
If from epoch is far future from current epoch then what will happen? I could not see this check? Should be a checking condition in here:
```go
        currentEpoch := helpers.SlotToEpoch(bs.GenesisTimeFetcher.CurrentSlot())
	if from > currentEpoch {
		return nil, status.Errorf(
			codes.InvalidArgument,
			errEpoch,
			currentEpoch,
			fromEpoch,
		)
	}
```

mxmar replied:
You're right. I've prepared this check as you has suggested, as it is in `ListValidatorAssignments` too. I've managed to reimplement this fragment too then:
```
minimalConsensusInfoRange, err := bs.MinimalConsensusInfoRange(bs.Ctx, requestedFromEpoch)
if err != nil {
	return status.Errorf(codes.Unavailable, "Could not get MinimalConsensusInfoRange for a stream: %v", err)
}
```

mxmar replied:
I've decide to move it to the upper level:
```
       currentEpoch := helpers.SlotToEpoch(bs.GenesisTimeFetcher.CurrentSlot())
	if requestedFromEpoch > currentEpoch {
		return status.Errorf(
			codes.InvalidArgument,
			errEpoch,
			currentEpoch,
			requestedFromEpoch,
		)
	}

	minimalConsensusInfoRange, err := bs.MinimalConsensusInfoRange(bs.Ctx, requestedFromEpoch)
	if err != nil {
		return status.Errorf(codes.Unavailable, "Could not get MinimalConsensusInfoRange for a stream: %v", err)
	}
```

mxmar committed:
Feature: 11 grpc van sub
- added new vanguard-apis;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11
- debugging;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- added sorting to minimalConsensusInfo;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- changes;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- added future minimal consensus info;
- moved minimal consensus info logic from blockchain to the rpc/beacon;
- only proper test is missing for minimal consensus info in rpc/beacon;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- rpc be modifications;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 11 grpc van sub
- test refactor;

Signed-off-by: mxmar <macht4tech@gmail.com>
